racket/collects/racket/private/dict.rkt

```diff
@@ -244,23 +196,39 @@
       (raise-argument-error 'dict-map "dict?" d))
     (proc (car x) (cdr x))))
 
+(define (assoc-fold-unique f init d)
+  (unless (assoc? d)
+    (raise-argument-error 'assoc-fold-unique "dict?" d))
+  (let loop ([xd d]
+             [acc init]
+             [seen (set)])
```

Comment:
Feel free to ignore this, but small lists (< ~100 elements) tend to perform a lot better than sets. It might be worthwhile to use one or the other based on the size of `d`.

Here's a microbenchmark that shows the difference: https://gist.github.com/Bogdanp/e67c9d1c8c33d6b0df2fbbb24b8c2e29

Reply:
I added hashes with `#true` keys to the microbenchmark:
https://gist.github.com/bennn/28cf78e1c832603b2fc8ba1c8f245c22

And it looks like both of those perform better than sets. So, I'd go with immutable hashes and change `(set-add ....)` to `(hash-set ... #true)`

Reply:
Hi @bennn and @Bogdanp. I've changed the code to use immtable hash sets.